zrml/prediction-markets/src/tests.rs

```diff
@@ -887,6 +887,55 @@ fn on_market_close_successfully_auto_closes_multiple_markets_after_stall() {
     });
 }
 
+#[test]
+fn market_close_manager_skips_the_genesis_block_with_timestamp_zero() {
+    // We ensure that a timestamp of zero will not be stored at genesis into LastTimeFrame storage.
+    let end: Moment = (5 * MILLISECS_PER_BLOCK).into();
+    ExtBuilder::default().build().execute_with(|| {
+        assert_ok!(PredictionMarkets::create_cpmm_market_and_deploy_assets(
+            Origin::signed(ALICE),
+            ALICE,
+            MarketPeriod::Timestamp(0..end),
+            gen_metadata(50),
+            MarketType::Categorical(3),
+            MarketDisputeMechanism::SimpleDisputes,
+            <Runtime as zrml_swaps::Config>::MinLiquidity::get(),
+            vec![<Runtime as zrml_swaps::Config>::MinWeight::get(); 4],
+        ));
+
+        let noop_mutation = |_: &crate::MarketIdOf<Runtime>,
+                             _: Market<
+            <Runtime as frame_system::Config>::AccountId,
+            <Runtime as frame_system::Config>::BlockNumber,
+            crate::MomentOf<Runtime>,
+        >|
+         -> DispatchResult { Ok(()) };
+
+        // Blocknumber = 0 and timestamp = 0
+        assert_eq!(Timestamp::get(), 0);
+        assert_ok!(PredictionMarkets::market_close_manager(0, noop_mutation));
+        assert_eq!(LastTimeFrame::<Runtime>::get(), None);
+
+        // Blocknumber != 0 and timestamp = 0
+        assert_eq!(Timestamp::get(), 0);
+        assert_ok!(PredictionMarkets::market_close_manager(1, noop_mutation));
+        assert_eq!(LastTimeFrame::<Runtime>::get(), None);
+
+        // Blocknumer = 0 and timestamp != 0
+        Timestamp::set_timestamp(end);
+        assert_ok!(PredictionMarkets::market_close_manager(0, noop_mutation));
+        assert_eq!(LastTimeFrame::<Runtime>::get(), None);
+
+        // Blocknumer != 0 and timestamp != 0
+        assert_ok!(PredictionMarkets::market_close_manager(1, noop_mutation));
```

Comment:
How is this check relevant? IIUC this returns Ok(()) from noop_mutation and all other cases above are returned from newly added condition check in market_close_manager right ?

Reply:
The first three checks ensure that no timestamp is stored at genesis or when it's zero. The blocknumber is also checked because the timestamp technically is undefined at genesis. The last check verifies that using a correct block number and timestamp results in storing the proper timestamp. The idea behind this is that a zero or undefined timestamp will never be stored within the storage `LastTimeFrame`, effectively avoiding block production stalling ranges from a zero/undefined timestamp to the next current valid timestamp within the range in: https://github.com/zeitgeistpm/zeitgeist/blob/6c0b74dc7078b51fd1426023a5174fb8058ee132/zrml/prediction-markets/src/lib.rs#L1946-L1955